### PR TITLE
💎 Ruby's Capstone Project: Catalog of My Things App

### PR DESCRIPTION
**Pull request by jicamargo**
# 💎 Ruby's Capstone Project: Catalog of My Things App

### authors:
 - @emarkees 
 - @MoyasiGinko 
 - @jicamargo 

### Description:
The "Catalog of My Things" app is a console-based application designed to help users manage collections of various items such as 📚 books, 🎵 music albums, 🎬 movies, and 🎮 games. The app employs a UML class diagram as its foundation and utilizes both 📄 JSON files and a 💽 database structure to store data.

### How to Build the "Catalog of My Things" App:
- Each class from the UML class diagram is implemented in separate `.rb` files.
- All 1️⃣-to-many relationships and parent-child aggregations between classes are correctly established.
- All properties outlined in the diagram are defined and initialized in the constructor method, with exceptions for 1️⃣-to-many relationship properties.
- Necessary methods, as specified in the diagram, are fully implemented:
  - `add_item` method in relevant classes:
    - Adds an instance of the Item class to the item collection.
    - Sets itself as a property of the item object using the appropriate setter.
  - `can_be_archived?` method in the Item class:
    - Returns true if the published_date is over 🔟 years old; otherwise, false.
  - `move_to_archive` method in the Item class:
    - Utilizes the `can_be_archived?` method.
    - Changes the archived property to true if `can_be_archived?` is true.
  - `can_be_archived?` methods in child classes (Book, MusicAlbum, Movie, Game):
    - Override the parent class method.
    - Return true based on specific criteria, utilizing parent's method if necessary.
- Comprehensive unit tests are written and executed for all implemented methods.

User Interface:
- A `main.rb` file serves as the entry point for the console app.
- At launch, the app presents users with a menu of options.
- Users can select an option and provide relevant parameters when needed.
- The app includes an option to quit.
- Available options include:
  - Listing all 📚 books, 🎵 music albums, 🎬 movies, and 🎮 games.
  - Listing genres, labels, authors, and sources.
  - Adding a 📚 book, 🎵 music album, 🎬 movie, or 🎮 game.
- Data is preserved by storing collections in 📄 JSON files.

Database:
- A `schema.sql` file is created with tables mirroring the class structure of the app.
- Due to the absence of inheritance in database tables, all properties and associations from the parent `Item` class are added as columns to child class tables.
- This approach ensures data integrity and facilitates future data storage and retrieval.

Additional Requirements:
- Code adheres to Ruby's best practices.
- Linter errors are resolved.
- GitFlow is properly followed for version control.
- Documentation is comprehensive and professional.

By implementing this "Catalog of My Things" app, users can effectively manage their collections and utilize both 📄 JSON files and a well-structured 💽 database for data storage and retrieval.

### CODE REVIEW

File: spec/class_album_spec.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+require_relative '../class_musicalbum_handler'
+require_relative '../class_genres_handler'
+
+RSpec.describe MusicalbumHandler do
+  let(:musicalbum_handler) { MusicalbumHandler.new }
```
Comment (Owusu-Desmond) [inline spec/class_album_spec.rb:5]: - [ ] Please make sure to add a test for the method `can_be_archived?()` in the MusicAlbum class. :pray:

File: spec/class_book_spec.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
+require_relative '../class_books_handler'
+require_relative '../class_genres_handler'
+require_relative '../class_labels_handler'
+require_relative '../class_authors_handler'
+
+RSpec.describe BookHandler do
+  let(:book_handler) { BookHandler.new }
```
Comment (Owusu-Desmond) [inline spec/class_book_spec.rb:7]: - [ ] Please make sure to add a test for the method `can_be_archived?()` in the Book class. :pray:

File: spec/class_author_spec.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
+require_relative '../class_authors_handler'
+require_relative '../class_author'
+
+RSpec.describe AuthorHandler do
+  let(:author_handler) { AuthorHandler.new }
```
Comment (Owusu-Desmond) [inline spec/class_author_spec.rb:5]: - [ ] Please make sure to add a test for the methods `can_be_archived?()` and `move_to_archived()` in the item class. :pray:

File: spec/class_game_spec.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
+require_relative '../class_games_handler'
+require_relative '../class_author'
+require_relative '../class_game'
+require_relative '../date_validator'
+
+RSpec.describe GameHandler do
```
Comment (Owusu-Desmond) [inline spec/class_game_spec.rb:6]: - [ ] Please make sure to add a test for the method `can_be_archived?()` in the Game class. :pray:

File: class_item.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
+require 'securerandom'
+
+# class_item.rb
+class Item
```
Comment (Owusu-Desmond) [inline class_item.rb:4]: - [ ] Please make you write a unit test for your item class and ensure to add a test for the methods `can_be_archived?()` and `move_to_archived()` in the item class. :pray:

**Review by Owusu-Desmond — CHANGES_REQUESTED** [review]
Hi @jicamargo, @emarkees and @MoyasiGinko  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights
📌 No Linters error ✅
📌 Descriptive pull request✅
📌 Great working on the set up of the project ✅
📌 Good commit messages ✅
📌 Professional README file ✅

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/class_album_spec.rb (hunk 1 above)
Comment (jicamargo) [reply spec/class_album_spec.rb:5]: thank you. @MoyasiGinko is goint to fixit

File: spec/class_book_spec.rb (hunk 2 above)
Comment (jicamargo) [reply spec/class_book_spec.rb:7]: thank you. @MoyasiGinko is goint to fixit

File: class_item.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  attr_reader :id, :archived
+  attr_accessor :genre, :author, :label, :publish_date
+
```
[the remark this answers is not in the document]
Comment (jicamargo) [reply class_item.rb:7]: thank you @Owusu-Desmond 
Changes Done!

File: README.md [hunk 7]
```diff
@@ -40,6 +37,9 @@
 - **Flexible and Expandable**: The modular design allows easy addition of new item types or enhancements to suit your needs.
 - **Efficient Search**: Search and retrieve items based on various criteria, enhancing the usability of your catalog.
 
+#### Demostration video <a name="live-demo"></a>
+  - [🚀 Watch the team talking about this project](https://drive.google.com/file/d/1toFii-z68Ov5vxcheXIpckndLZ1bIaDO/view?usp=sharing)
+
```
[the remark this answers is not in the document]
Comment (jicamargo) [reply README.md:42]: thank you @Owusu-Desmond .
Changes done!. In the readme.md there is a new demo video link 👍🏽

File: class_item.rb (hunk 5 above)
Comment (jicamargo) [reply class_item.rb:4]: thank you @Owusu-Desmond 
We add a new test file: class_item_spec.rb with the tests required. ✔️👍🏽

File: spec/class_game_spec.rb (hunk 4 above)
Comment (jicamargo) [reply spec/class_game_spec.rb:6]: Thank you. I add. class_game_can_be_archive_spec.rb

File: spec/class_album_spec.rb (hunk 1 above)
Comment (jicamargo) [reply spec/class_album_spec.rb:5]: Thank you. class_album_can_be_archived_spec.rb added! 👍🏽

File: spec/class_author_spec.rb (hunk 3 above)
Comment (jicamargo) [reply spec/class_author_spec.rb:5]: Thank you. I added class_item_spec.rb

File: README.md [hunk 8]
```diff
@@ -40,6 +37,9 @@
 - **Flexible and Expandable**: The modular design allows easy addition of new item types or enhancements to suit your needs.
 - **Efficient Search**: Search and retrieve items based on various criteria, enhancing the usability of your catalog.
 
+#### Demostration video <a name="live-demo"></a>
+  - [🚀 Watch the team talking about this project](https://drive.google.com/file/d/1J8v6inrGrJsUpXXhJ0QO91eOwAnqpa0l/view?usp=sharing)
```
Comment (gemmen29) [inline README.md:41]: - [ ] Kindly keep your video with a maximum length of 5 minutes as per this [requirement](https://github.com/microverseinc/curriculum-ruby/blob/main/group-capstone/catalog_of_my_things.md#:~:text=It%20is%20a%20video%20with%20a%20maximum%20length%20of%205%20minutes.).

**Review by gemmen29 — CHANGES_REQUESTED** [review]
Hi Team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Highlights💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 8 above)
Comment (jicamargo) [reply README.md]: Thank you. 💫💥
Readme updated with new Video link.
- do you Want to know what the authors are most proud of? 🚀 Watch the video, specially at 3:21 minute 👀.

**Review by mwanawabangona — APPROVED** [review]
Hi @Team!

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:
Congratulations!  


## Highlights
*   No linter errors :heavy_check_mark:
*   Repo looks professional :briefcase:
*   Solid pull request scroll
*   Good quality commits
*   Good job Implementing the requested changes!! :clap: :clap: :clap:

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**